server/controllers/finance/reports/balance/index.js

```diff
@@ -0,0 +1,225 @@
+  session.enterprise = req.session.enterprise;
+
+  _.defaults(params, { orientation : 'landscape' ,user : req.session.user });
+
+  try {
+    report = new ReportManager(TEMPLATE, req.session, params);
+  } catch (e) {
+    return next(e);
+  }
+
+  let accounts;
+  let totals;
+
+  params.enterpriseId = session.enterprise.id;
+
+  balanceReporting(params)
+    .then(balances => {
+      return processAccounts(balances, accounts, totals);
+    })
+    .then(result => {
+      data = { accounts: result.accounts, totals: result.totals, session };
+      return report.render(data);
+    })
+    .then(result => {
+      res.set(result.headers).send(result.report);
+    })
+    .catch(next)
+    .done();
+}
+
+/**
+ * @method processAccounts
+ * @description process and format accounts balance
+ * @param {object} balances The result of balanceReporting function
+ */
+function processAccounts(balances, accounts, totals) {
+
+  // format and process opening balance for accounts
+  accounts = balances.beginning.reduce(function (init, row) {
+    var account = init, id = row.number;
+    var obj = account[id] = {}, sold = getSold(row);
+    obj.label = row.label;
+    obj.beginDebit = sold.debit;
+    obj.beginCredit = sold.credit;
+    obj.middleDebit = 0;
+    obj.middleCredit = 0;
+    obj.is_charge = row.is_charge;
+    obj.is_asset = row.is_asset;
+    account[id] = obj;
+    return account;
+  }, {});
+
+  // format and process the monthly balance for accounts
+  balances.middle.forEach(function (row){
+    var account = accounts[row.number] || {};
+    account.middleDebit = row.debit;
+    account.middleCredit = row.credit;
+    account.label = row.label;
+    account.is_charge = row.is_charge;
+    account.is_asset = row.is_asset;
+    accounts[row.number] = account;
+  });
+
+  Object.keys(accounts).forEach(function (item){
+    accounts[item].endDebit = 0;
+    accounts[item].endCredit = 0;
+    var sold = (accounts[item].beginDebit || 0 - accounts[item].beginCredit || 0) + (accounts[item].middleDebit - accounts[item].middleCredit);
+    if(sold < 0){
+      accounts[item].endCredit = sold * -1;
+    }else{
+     accounts[item].endDebit = sold;
+    }
+  });
+
+  // process for getting totals
+  totals = Object.keys(accounts).reduce(function (totals, key) {
+    var account = accounts[key];
+    totals.beginDebit += (account.beginDebit || 0);
+    totals.beginCredit += (account.beginCredit || 0);
+    totals.middleDebit += (account.middleDebit || 0);
+    totals.middleCredit += (account.middleCredit || 0);
+    totals.endDebit += (account.endDebit || 0);
+    totals.endCredit += (account.endCredit || 0);
+    return totals;
+  }, {
+    beginDebit : 0,
+    beginCredit : 0,
+    middleDebit: 0,
+    middleCredit : 0,
+    endDebit : 0,
+    endCredit : 0
+  });
+
+  return { accounts, totals };
+}
+
+/**
+ * @function getSold
+ * @description return the sold of an account
+ * @param {object} account An objec
+ */
+function getSold (item){
```

Comment:
Wow!  This is a complex function.   It could use some comments.

This is a non-blocking comment.